### PR TITLE
add east det model

### PR DESCRIPTION
**Pull request by tonytonglt**
add east det model, where the best f-score is 84.95%, compared with modelzoo +3.99%, compared with paddleocr +0.05%.

### CODE REVIEW

File: mindocr/data/transforms/east_transforms.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
+import math
+import numpy as np
+import cv2
+from PIL import Image
+from shapely.geometry import Polygon
+
+__all__ = ['EASTProcessTrain']
+
+
+def cal_distance(x1, y1, x2, y2):
```
Comment (HaoyangLee) [inline mindocr/data/transforms/east_transforms.py:10]: Some potential refactoring on these functions to make them more clear and readable.
- Functions that are only called by other functions, i.e., internal use, can be defined as 'private' function like `_func()`
- Seems all functions outside the class can be defined inside the class.

File: mindocr/data/transforms/east_transforms.py
Comment (SamitHuang) [inline mindocr/data/transforms/east_transforms.py:1]: rename the file to `det_east_transform.py`

**Review by HaoyangLee — APPROVED** [review]

File: configs/det/east/east_r50_icdar15.yaml [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  loss_scale: 512
+  scale_factor: 2
+  scale_window: 1000
+
+train:
+  ckpt_save_dir: './tmp_det'
+  dataset_sink_mode: True
+  dataset:
+    type: DetDataset
+    dataset_root: /home/tongli/data/icdar2015/
```
Comment (hadipash) [inline configs/det/east/east_r50_icdar15.yaml:61]: Let's not use person specific folders.

File: mindocr/data/transforms/east_transforms.py [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+
+class EASTProcessTrain:
+    def __init__(self, scale=0.25, length=512):
+        super(EASTProcessTrain, self).__init__()
+        self.scale = scale
+        self.length = length
+
+    def __call__(self, data):
+        vertices, labels = extract_vertices(data['label'])
+        img = Image.open(data['img_path'])
```
Comment (hadipash) [inline mindocr/data/transforms/east_transforms.py:413]: Why is the image loaded here? Shouldn't it be inside data loader?

File: mindocr/data/transforms/east_transforms.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+        img, vertices = adjust_height(img, vertices)
+        img, vertices = rotate_img(img, vertices)
+        img, vertices = crop_img(img, vertices, labels, self.length)
```
Comment (hadipash) [inline mindocr/data/transforms/east_transforms.py:416]: I think these augmentations should be extracted into individual operations and described inside the `transform_pipeline` of the config file.

File: mindocr/losses/det_loss.py [hunk 5]
```diff
@@ -149,3 +149,69 @@ def construct(self, pred, gt, mask):
 
         return (pos_loss.sum() + neg_loss.sum()) / \
                (pos_count.astype(ms.float32).sum() + neg_count.astype(ms.float32).sum() + self._eps)
+
+
+class DiceCoefficient(nn.Cell):
```
[the remark this answers is not in the document]
Comment (hadipash) [reply mindocr/losses/det_loss.py:154]: I think is possible. Just set mask to `None` by default.

File: mindocr/losses/det_loss.py [hunk 6]
```diff
@@ -149,3 +149,69 @@ def construct(self, pred, gt, mask):
...
+
+    def construct(self, true_cls, pred_cls):
+        intersection = self.sum(true_cls * pred_cls, ())
+        union = self.sum(true_cls, ()) + self.sum(pred_cls, ()) + self.eps
+        loss = 1. - (2 * intersection / union)
+
+        return loss
+
+
+class MyMin(nn.Cell):
```
Comment (hadipash) [inline mindocr/losses/det_loss.py:168]:
```suggestion
class MyMin(nn.LossBase):
```

File: mindocr/losses/det_loss.py [hunk 7]
```diff
@@ -149,3 +149,69 @@ def construct(self, pred, gt, mask):
...
+    def __init__(self):
+        super(MyMin, self).__init__()
+        self.abs = ops.Abs()
+
+    def construct(self, a, b):
+        return (a + b - self.abs(a - b)) / 2
+
+
+class EastLoss(nn.Cell):
```
Comment (hadipash) [inline mindocr/losses/det_loss.py:177]:
```suggestion
class EastLoss(nn.LossBase):
```

File: mindocr/losses/det_loss.py [hunk 8]
```diff
@@ -149,3 +149,69 @@ def construct(self, pred, gt, mask):
...
+        self.eps = 1e-5
+        self.dice = DiceCoefficient()
+
+    def construct(
+            self,
+            pred,
+            score_map,
+            geo_map,
+            training_mask):
+        ans = self.sum(score_map)
```
Comment (hadipash) [inline mindocr/losses/det_loss.py:195]:
```suggestion
ans = score_map.sum()
```

File: mindocr/models/heads/det_east_head.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+        super(EASTHead, self).__init__()
+        self.conv1 = nn.Conv2d(in_channels, 1, 1)
+        self.sigmoid1 = nn.Sigmoid()
+        self.conv2 = nn.Conv2d(in_channels, 4, 1)
+        self.sigmoid2 = nn.Sigmoid()
+        self.conv3 = nn.Conv2d(in_channels, 1, 1)
+        self.sigmoid3 = nn.Sigmoid()
+        self.scope = scope
+        self.concat = ops.Concat(axis=1)
+        self.PI = 3.1415926535898
```
Comment (hadipash) [inline mindocr/models/heads/det_east_head.py:15]:
```suggestion
self.PI = math.pi
```

File: mindocr/data/transforms/east_transforms.py (hunk 4 above)
Comment (tonytonglt) [reply mindocr/data/transforms/east_transforms.py:416]: Yes, I think they could be. But it seems that these augmentations are only used in EAST, maybe they could also be integrated in EASTProcessTrain class together? Like PaddleOCR does: https://github.com/PaddlePaddle/PaddleOCR/blob/013870d9bc963f7508d5a76f81da872b84471a71/ppocr/data/imaug/east_process.py

Comment (hadipash) [reply mindocr/data/transforms/east_transforms.py:416]: From what I see, some operations are very aligned with the DBNet augmentations: [rotate](https://github.com/zhtmike/mindocr/blob/9d959ae0b18a047bc76213db180b68742f1bcfcd/mindocr/data/transforms/general_transforms.py#L177) (work in progress: not integrated yet), [scale](https://github.com/mindspore-lab/mindocr/blob/d2163929222e10b312c12719b6d10b157b28dbfb/mindocr/data/transforms/general_transforms.py#L193) / resize, [crop](https://github.com/mindspore-lab/mindocr/blob/d2163929222e10b312c12719b6d10b157b28dbfb/mindocr/data/transforms/general_transforms.py#L222). Maybe you can adjust existing transformations to fit your needs?

Also, we need to have an ability to control each augmentation's probability. For example, applying rotation too frequently on some datasets will lead to degraded performance (e.g. datasets with straight (not tilted) paper forms). And to have a better control of each augmentation, it's better to separate them.

File: mindocr/losses/det_loss.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tonytonglt) [reply mindocr/losses/det_loss.py:154]: I've tried to reuse the DiceLoss, but I found there's slightly f-score decrease, so I‘m still using the DiceCoefficient.

File: mindocr/data/transforms/east_transforms.py (hunk 1 above)
Comment (tonytonglt) [reply mindocr/data/transforms/east_transforms.py]: Modified.

File: configs/det/east/east_r50_icdar15.yaml (hunk 2 above)
Comment (tonytonglt) [reply configs/det/east/east_r50_icdar15.yaml]: Modified.

File: mindocr/data/transforms/east_transforms.py (hunk 3 above)
Comment (tonytonglt) [reply mindocr/data/transforms/east_transforms.py]: Modified, using data['image'] from dataloader instead.

File: mindocr/losses/det_loss.py (hunk 6 above)
Comment (tonytonglt) [reply mindocr/losses/det_loss.py]: Modified to a private function.

File: mindocr/models/heads/det_east_head.py (hunk 9 above)
Comment (tonytonglt) [reply mindocr/models/heads/det_east_head.py]: Modified

File: mindocr/losses/det_loss.py (hunk 7 above)
Comment (tonytonglt) [reply mindocr/losses/det_loss.py]: Modified.

File: configs/det/east/east_r50_icdar15.yaml
[the remark this answers is not in the document]
Comment (tonytonglt) [reply configs/det/east/east_r50_icdar15.yaml:1]: Done.

File: mindocr/data/transforms/east_transforms.py
Comment (tonytonglt) [reply mindocr/data/transforms/east_transforms.py:1]: Done.

**Review by SamitHuang — APPROVED** [review]

File: mindocr/postprocess/det_postprocess.py [hunk 10]
```diff
@@ -143,3 +145,101 @@ def _calc_score(pred, mask, contour):
...
+        valid_geo = geo[:, xy_text[:, 0], xy_text[:, 1]]  # 5 x n
+        polys_restored, index = self.restore_polys(valid_pos, valid_geo, score.shape)
+        if polys_restored.size == 0:
+            return [(np.array([[[1, 2], [3, 4], [5, 6], [7, 8]]], 'float32'), np.array(0.))]
+
+        boxes = np.zeros((polys_restored.shape[0], 9), dtype=np.float32)
+        boxes[:, :8] = polys_restored
+        boxes[:, 8] = score[xy_text[index, 0], xy_text[index, 1]]
+        boxes = lanms.merge_quadrangle_n9(boxes.astype('float32'), self.nms_thresh)
+        return [(boxes[:, :8].reshape(-1, 4, 2), boxes[:, 8])]
```
Comment (SamitHuang) [inline mindocr/postprocess/det_postprocess.py:184]: 后处理的输出格式统一用字典，具体字段命名，请参考 https://github.com/mindspore-lab/mindocr/pull/309

**Comment by tonytonglt** [comment]
> Please also update Supported Algorithm Section in Readme.md and reamd_cn.md



> 
I've already pushed the readme_cn.md, but it seems does not appear in this review page. 
Please see:
https://github.com/mindspore-lab/mindocr/pull/264/commits/04bd76c4d1ea939149baa044577d3d130d9f33ac#diff-a2f5307adeaccc5310fc03f22dbe74a5222dfb2a1344e2c05c5b02c1004b792dR1

File: mindocr/postprocess/det_postprocess.py (hunk 10 above)
Comment (tonytonglt) [reply mindocr/postprocess/det_postprocess.py]: Done. Please check here:
https://github.com/mindspore-lab/mindocr/pull/264/commits/c7c8cb6ca332a96132a08e08c6e6b1d82daf8f4b#diff-21c18c43684dc9711f08f458b437073639a0acf7ddcd0a9a204352cd83ce36d9R190

**Comment by SamitHuang** [comment]
refer to: https://github.com/mindspore-lab/mindocr/pull/316

File: configs/det/east/east_r50_icdar15.yaml [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+          target_size: [720, 1280]
+          force_divisable: False
+      - NormalizeImage:
+          bgr_to_rgb: False
+          is_hwc: True
+          mean: imagenet
+          std: imagenet
+      - ToCHWImage:
+    #  the order of the dataloader list, matching the network input and the labels for evaluation
+    output_columns: [ 'image', 'polys', 'ignore_tags' ]
```
[the remark this answers is not in the document]
Comment (tonytonglt) [reply configs/det/east/east_r50_icdar15.yaml]: fixed

File: mindocr/postprocess/det_postprocess.py [hunk 12]
```diff
@@ -290,3 +292,135 @@ def _generate_box(self, score, label, shape):
...
+
+class EASTPostprocess:
+    def __init__(self, score_thresh=0.8, nms_thresh=0.2, rescale_fields=None):
+        self.score_thresh = score_thresh
+        self.nms_thresh = nms_thresh
+        if rescale_fields is None:
+            rescale_fields = []
+        self.rescale_fields = rescale_fields
+
+    def __call__(self, pred, shape_list=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (tonytonglt) [reply mindocr/postprocess/det_postprocess.py]: fixed

[the remark this answers is not in the document]
Comment (tonytonglt) [reply mindocr/postprocess/det_postprocess.py]: fixed

File: mindocr/losses/det_loss.py [hunk 13]
```diff
@@ -5,7 +5,7 @@
 from mindspore import Tensor
 import mindspore.numpy as mnp
 
-__all__ = ['L1BalancedCELoss', 'PSEDiceLoss']
+__all__ = ['L1BalancedCELoss', 'PSEDiceLoss', 'EastLoss']
```
Comment (SamitHuang) [inline mindocr/losses/det_loss.py:8]: Let's keep the naming convention: EastLoss -> EASTLoss

File: deploy/py_infer/src/data_process/postprocess/east_postprocess.py [hunk 14]
```diff
@@ -0,0 +1,126 @@
+import math
+import numpy as np
+import lanms
+from mindspore import Tensor
+
+
+class EASTPostprocess:
```
Comment (SamitHuang) [inline deploy/py_infer/src/data_process/postprocess/east_postprocess.py:7]: 这部分代码需要跟train的更新同步一下？

**Review by SamitHuang — APPROVED** [review]
